source/common/upstream/upstream_impl.h

```diff
@@ -103,8 +104,16 @@ class HostDescriptionImpl : virtual public HostDescription,
     return metadata_;
   }
   void metadata(MetadataConstSharedPtr new_metadata) override {
-    absl::WriterMutexLock lock(&metadata_mutex_);
-    metadata_ = new_metadata;
+    auto& new_socket_factory = resolveTransportSocketFactory(address_, new_metadata.get());
+    {
+      absl::WriterMutexLock lock(&metadata_mutex_);
+      metadata_ = new_metadata;
+    }
+    {
+      absl::WriterMutexLock lock(&socket_factory_mutex_);
```

Comment:
is it semantically sound to have a separate lock for the socket factory? If there's a thread wakeup between these two locks will there be an inconsistent view of the data?

Reply:
Simplified it and one mutex is used to protect metadata_ and fields dependent on it - in this case socket_factory_.